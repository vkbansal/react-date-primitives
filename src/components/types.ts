// tslint:disable:max-line-length
/**
 * A Day of the month
 */
export interface DayOfMonth {
    /**
     * The date
     */
    date: Date;
    /**
     * Is the `date` in current month?
     */
    inCurrentMonth: boolean;
    /**
     * Is the `date` between `startDate` and `endDate`?
     * By default this value is `false`.
     * This will be populated correctly only when both `startDate` and `endDate` are defined and are valid.
     */
    inRange: boolean;
    /**
     * Is the date selected?
     */
    selected: boolean;
    /**
     * Is the date disabled?
     * By default this value is `false`.
     * It will be `true` for all the dates before `minDate` and after `maxDate`.
     */
    disabled: boolean;
}

/**
 * Dropdown option
 */
export interface CalendarDropdownOption {
    /**
     * Value of the dropdown
     */
    value: number;
    /**
     * Is the value selected?
     */
    selected: boolean;
    /**
     * Is the value disabled?
     */
    disabled: boolean;
}

/**
 * CalendarMonth `render` prop is called with this object
 */
export interface CalendarMonthRenderProps {
    /**
     * Days of current month.
     *
     * @see DayOfMonth
     */
    days: DayOfMonth[][];
    /**
     * Current month.
     */
    month: Date;
    /**
     * Values for creating month dropdown.
     * Months start from `0`, similar to [`Date.proptotype.getMonth()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getMonth).
     * By default, this value is `undefined`. To populate it, set `showDropdowns` prop as `true`.
     *
     * @see CalendarDropdownOption
     */
    monthsDropdown?: CalendarDropdownOption[];
    /**
     * Values for creating year dropdown.
     * By default, this value is `undefined`. To populate it, set `showDropdowns` prop as `true`.
     *
     * @see CalendarDropdownOption
     */
    yearsDropdown?: CalendarDropdownOption[];
}

/**
 * Props required by <CalendarMonth /> component
 */
export interface CalendarMonthProps {
    /**
     * Current month to be shown
     */
    month: Date;
    /**
     * The initially selected date when using as a single date-picker.
     * The start of the initially selected date range when used as a daterange-picker.
     */
    startDate?: Date;
    /**
     * The end of the initially selected date range (only required when used as a daterange-picker)
     */
    endDate?: Date;
    /**
     * The earliest date a user may select
     */
    minDate?: Date;
    /**
     * The latest date a user may select
     */
    maxDate?: Date;
    /**
     * If set as `true`, then `monthsDropdown` and `yearsDropdown` are populated in the argument given to `render` prop.
     */
    showDropdowns?: boolean;
    /**
     * The main function, which be used for rendering.
     *
     * @see CalendarMonthRenderProps
     */
    render(props: CalendarMonthRenderProps): React.ReactNode;
}

/**
 * A date range
 */
export interface DateRange {
    /**
     * Starting date of the range
     */
    startDate: Date;
    /**
     * End date of the range
     */
    endDate?: Date;
}

/**
 * @private
 */
export type PickedCalendarMonthProps = Pick<
    CalendarMonthProps,
    'startDate' | 'endDate' | 'minDate' | 'maxDate'
>;

/**
 * DateRangeControl `render` prop is called with this object
 *
 * @extends PickedCalendarMonthProps
 */
export interface DateRangeControlRenderProps extends PickedCalendarMonthProps {
    /**
     * Callback for handling click on a date
     */
    onDayClick(date: Date): void;
    /**
     * Callback for handling mouse hover on a date
     */
    onDayHover(date: Date): void;
    /**
     * The months to be shown to the user
     */
    months: Date[];
    /**
     * Callback for moving to next month
     */
    moveForward(): void;
    /**
     * Callback for moving to prev month
     */
    moveBackward(): void;
}

/**
 * @extends PickedCalendarMonthProps
 */
export interface DateRangeControlProps extends PickedCalendarMonthProps {
    /**
     * Callback when start and/or end dates are changed
     *
     * @see DateRange
     */
    onDatesChange?(dates: DateRange): void;
    /**
     * By default 2 consecutive months will be shown,
     * but if you want to show 3 months instead, set this to `true`
     */
    showThreeMonths?: boolean;
    /**
     * The main function, which be used for rendering.
     * It is called with an object.
     *
     * @see DateRangeControlRenderProps
     */
    render(props: DateRangeControlRenderProps): React.ReactNode;
}

/**
 * @private
 */
export interface DateRangeControlState {
    startDate?: Date;
    endDate?: Date;
    months: Date[];
    selectionActive: boolean;
}

/**
 * @private
 */
export type DayOfMonthSubset = Pick<DayOfMonth, 'date' | 'inCurrentMonth'>;
