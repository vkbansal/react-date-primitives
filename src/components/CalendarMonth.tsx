import * as React from 'react';

import {
    startOfMonth,
    isDayAfter,
    isDayBefore,
    isSameDay,
    setMonth,
    getDaysOfMonth
} from '../utils';

import { CalendarMonthProps, DayOfMonth, CalendarDropdownOption } from './types';

/**
 * Primitive react component that can be used to make a datepicker.
 */
export class CalendarMonth extends React.Component<CalendarMonthProps> {
    private getDaysofMonth(props: CalendarMonthProps): DayOfMonth[][] {
        const { month, startDate, endDate, minDate, maxDate } = props;
        const currentMonth = getDaysOfMonth(month);

        const startDateDisabled = !!startDate && !!minDate && !isDayAfter(startDate, minDate);

        return currentMonth.map(
            (week): DayOfMonth[] =>
                week.map(
                    (day): DayOfMonth => {
                        const { date, inCurrentMonth } = day;
                        const disabled =
                            !inCurrentMonth ||
                            Boolean(
                                (minDate && isDayBefore(date, minDate)) ||
                                    (maxDate && isDayAfter(date, maxDate))
                            );

                        const selected =
                            !disabled &&
                            Boolean(
                                (startDate && isSameDay(date, startDate)) ||
                                    (endDate && isSameDay(date, endDate))
                            );

                        const inRange = Boolean(
                            startDate &&
                                isDayAfter(date, startDate) &&
                                endDate &&
                                isDayBefore(date, endDate)
                        );

                        return {
                            date,
                            inRange:
                                inCurrentMonth &&
                                !startDateDisabled &&
                                !disabled &&
                                (selected || inRange),
                            selected,
                            disabled,
                            inCurrentMonth
                        };
                    }
                )
        );
    }

    private getDropdowns = (
        props: CalendarMonthProps
    ): [CalendarDropdownOption[], CalendarDropdownOption[]] => {
        const { minDate, maxDate, month } = props;

        const firstDay = startOfMonth(month);
        const today = new Date();
        const minYear = minDate ? minDate.getFullYear() : today.getFullYear() - 50;
        const maxYear = maxDate ? maxDate.getFullYear() + 1 : today.getFullYear() + 5;

        const dropdownMonths: CalendarDropdownOption[] = Array.from({ length: 12 }, (_, i) => ({
            disabled: Boolean(
                (minDate && isDayBefore(setMonth(month, i), minDate)) ||
                    (maxDate && isDayAfter(setMonth(month, i), maxDate))
            ),
            value: i,
            selected: firstDay.getMonth() === i
        }));

        const dropdownYears: CalendarDropdownOption[] = Array.from(
            { length: maxYear - minYear },
            (_, i) => ({
                value: i + minYear,
                disabled: false,
                selected: i + minYear === firstDay.getFullYear()
            })
        );

        return [dropdownMonths, dropdownYears];
    };

    render() {
        const { month, render, showDropdowns } = this.props;

        let monthsDropdown: CalendarDropdownOption[] | undefined;
        let yearsDropdown: CalendarDropdownOption[] | undefined;

        if (showDropdowns) {
            [monthsDropdown, yearsDropdown] = this.getDropdowns(this.props);
        }

        return render({
            days: this.getDaysofMonth(this.props),
            month: startOfMonth(month),
            monthsDropdown,
            yearsDropdown
        });
    }
}
