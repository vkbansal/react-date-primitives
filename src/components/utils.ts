import { DayOfMonthSubset } from './types';

const monthsCache: Map<string, DayOfMonthSubset[][]> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDatesofMonth(month: Date): DayOfMonthSubset[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DayOfMonthSubset[][] = Array.from(
        { length: 6 },
        (_1, i): DayOfMonthSubset[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonthSubset => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
