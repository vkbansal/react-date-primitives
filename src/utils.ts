export interface DayOfMonth {
    date: Date;
    inCurrentMonth: boolean;
}

export type DaysOfMonth = DayOfMonth[][];

const monthsCache: Map<string, DaysOfMonth> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDaysOfMonth(month: Date): DayOfMonth[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DaysOfMonth = Array.from(
        { length: 6 },
        (_1, i): DayOfMonth[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonth => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
