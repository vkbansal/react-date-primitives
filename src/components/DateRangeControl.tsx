import * as React from 'react';

import { startOfMonth, addMonths, isDayBefore, callIfExists } from '../utils';
import { DateRangeControlProps, DateRangeControlState } from './types';

/**
 * Controller component for making a date-range picker in conjunction with [CalendarMonth]()
 */
export class DateRangeControl extends React.Component<
    DateRangeControlProps,
    DateRangeControlState
> {
    private numOfMonths: 2 | 3;

    constructor(props: DateRangeControlProps) {
        super(props);

        const { startDate, endDate, showThreeMonths } = props;

        this.numOfMonths = showThreeMonths ? 3 : 2;
        const thisMonth = startOfMonth(new Date());

        this.state = {
            startDate,
            endDate,
            months: Array.from({ length: this.numOfMonths }, (_, i) =>
                addMonths(startDate || thisMonth, i)
            ),
            selectionActive: false
        };
    }

    componentWillReceiveProps(nextProps: DateRangeControlProps) {
        const { startDate, endDate } = nextProps;

        this.setState((state) => {
            return {
                ...state,
                startDate,
                endDate,
                months: Array.from({ length: this.numOfMonths }, (_, i) =>
                    addMonths(startDate || state.months[0], i)
                )
            };
        });
    }

    handleDayClick = (day: Date) => {
        this.setState((state) => {
            const { startDate, selectionActive } = state;

            let newState: Pick<DateRangeControlProps, 'startDate' | 'endDate'>;

            if (startDate && selectionActive && !isDayBefore(day, startDate)) {
                newState = {
                    startDate,
                    endDate: day
                };

                callIfExists(this.props.onDatesChange, newState);

                return {
                    selectionActive: false,
                    ...newState
                };
            }

            newState = {
                startDate: day,
                endDate: undefined
            };

            callIfExists(this.props.onDatesChange, newState);

            return {
                selectionActive: true,
                ...newState
            };
        });
    };

    handleDayHover = (day: Date) => {
        const { selectionActive, startDate } = this.state;

        if (selectionActive && startDate && !isDayBefore(day, startDate)) {
            this.setState(() => ({
                endDate: day
            }));
        }
    };

    handleNavigation = (months: number) => () => {
        this.setState((state) => {
            return {
                months: state.months.map((month) => addMonths(month, months))
            };
        });
    };

    render() {
        const { render, minDate, maxDate } = this.props;
        const { months, startDate, endDate } = this.state;

        return render({
            months,
            startDate,
            endDate,
            minDate,
            maxDate,
            onDayClick: this.handleDayClick,
            onDayHover: this.handleDayHover,
            moveForward: this.handleNavigation(1),
            moveBackward: this.handleNavigation(-1)
        });
    }
}
