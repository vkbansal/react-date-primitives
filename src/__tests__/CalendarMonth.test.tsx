import * as React from 'react';
import { shallow } from 'enzyme';

import { CalendarMonth, CalendarMonthRenderProps } from '../';
import { addDays, addMonths } from '../components/utils';

describe('<CalenderMonth /> tests', () => {
    const month = new Date(Date.UTC(2017, 0 /* Jan */, 1, 0, 0, 0, 0));

    test('calls render props with given month and its days', () => {
        const render = jest.fn();
        shallow(<CalendarMonth month={month} render={render} />);

        expect(render).toHaveBeenCalledTimes(1);
        const args: CalendarMonthRenderProps = render.mock.calls[0][0];
        expect(args.days).toMatchSnapshot();
        expect(args.monthsDropdown).not.toBeDefined();
        expect(args.yearsDropdown).not.toBeDefined();
    });

    test('render prop has proper days selected when given', () => {
        const render = jest.fn();
        shallow(
            <CalendarMonth
                month={month}
                minDate={addDays(month, 5)}
                startDate={addDays(month, 10)}
                endDate={addDays(month, 20)}
                maxDate={addDays(month, 25)}
                render={render}
            />
        );

        const args: CalendarMonthRenderProps = render.mock.calls[0][0];
        // srd in snapshot means selected, inRange, disabled
        expect(args.days).toMatchSnapshot();
    });

    test('minDate and maxDate get priorty over startDate and endDate', () => {
        const render = jest.fn();
        shallow(
            <CalendarMonth
                month={month}
                minDate={addDays(month, 10)}
                startDate={addDays(month, 5)}
                endDate={addDays(month, 25)}
                maxDate={addDays(month, 20)}
                render={render}
            />
        );

        const args: CalendarMonthRenderProps = render.mock.calls[0][0];
        // srd in snapshot means selected, inRange, disabled
        expect(args.days).toMatchSnapshot();
    });

    test('showDropdowns gives apt props', () => {
        const render = jest.fn();
        shallow(<CalendarMonth showDropdowns month={month} render={render} />);

        const args: CalendarMonthRenderProps = render.mock.calls[0][0];
        expect(args.monthsDropdown).toMatchSnapshot();
        expect(args.yearsDropdown).toMatchSnapshot();
    });

    test('dropdowns are updated when minDate, maxDate are given', () => {
        const render = jest.fn();
        const component = shallow(<CalendarMonth showDropdowns month={month} render={render} />);

        let args: CalendarMonthRenderProps = render.mock.calls[0][0];
        expect(args.monthsDropdown).toMatchSnapshot();
        expect(args.yearsDropdown).toMatchSnapshot();

        component.setProps({
            minDate: addMonths(month, 2),
            maxDate: addMonths(month, 10)
        });

        args = render.mock.calls[1][0];
        expect(args.monthsDropdown[1]).toMatchObject({
            disabled: true
        });

        expect(args.monthsDropdown[11]).toMatchObject({
            disabled: true
        });
        expect(args.yearsDropdown.length).toBe(1);

        expect(args.monthsDropdown).toMatchSnapshot();
        expect(args.yearsDropdown).toMatchSnapshot();
    });
});
