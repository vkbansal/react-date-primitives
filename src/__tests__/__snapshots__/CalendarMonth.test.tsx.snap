// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalenderMonth /> tests calls render props with given month and its days 1`] = `
------------------------------------------------------------------------------
|                                January 2017                                |
------------------------------------------------------------------------------
| 01 (---) | 02 (---) | 03 (---) | 04 (---) | 05 (---) | 06 (---) | 07 (---) |
------------------------------------------------------------------------------
| 08 (---) | 09 (---) | 10 (---) | 11 (---) | 12 (---) | 13 (---) | 14 (---) |
------------------------------------------------------------------------------
| 15 (---) | 16 (---) | 17 (---) | 18 (---) | 19 (---) | 20 (---) | 21 (---) |
------------------------------------------------------------------------------
| 22 (---) | 23 (---) | 24 (---) | 25 (---) | 26 (---) | 27 (---) | 28 (---) |
------------------------------------------------------------------------------
| 29 (---) | 30 (---) | 31 (---) | 01 (--d) | 02 (--d) | 03 (--d) | 04 (--d) |
------------------------------------------------------------------------------
| 05 (--d) | 06 (--d) | 07 (--d) | 08 (--d) | 09 (--d) | 10 (--d) | 11 (--d) |
------------------------------------------------------------------------------
`;

exports[`<CalenderMonth /> tests dropdowns are updated when minDate, maxDate are given 1`] = `
----------
0 (--)
----------
1 (--)
----------
2 (--)
----------
3 (--)
----------
4 (--)
----------
5 (--)
----------
6 (--)
----------
7 (--)
----------
8 (--)
----------
9 (--)
----------
10 (--)
----------
11 (--)
----------
`;

exports[`<CalenderMonth /> tests dropdowns are updated when minDate, maxDate are given 2`] = `
----------
1969 (--)
----------
1970 (--)
----------
1971 (--)
----------
1972 (--)
----------
1973 (--)
----------
1974 (--)
----------
1975 (--)
----------
1976 (--)
----------
1977 (--)
----------
1978 (--)
----------
1979 (--)
----------
1980 (--)
----------
1981 (--)
----------
1982 (--)
----------
1983 (--)
----------
1984 (--)
----------
1985 (--)
----------
1986 (--)
----------
1987 (--)
----------
1988 (--)
----------
1989 (--)
----------
1990 (--)
----------
1991 (--)
----------
1992 (--)
----------
1993 (--)
----------
1994 (--)
----------
1995 (--)
----------
1996 (--)
----------
1997 (--)
----------
1998 (--)
----------
1999 (--)
----------
2000 (--)
----------
2001 (--)
----------
2002 (--)
----------
2003 (--)
----------
2004 (--)
----------
2005 (--)
----------
2006 (--)
----------
2007 (--)
----------
2008 (--)
----------
2009 (--)
----------
2010 (--)
----------
2011 (--)
----------
2012 (--)
----------
2013 (--)
----------
2014 (--)
----------
2015 (--)
----------
2016 (--)
----------
2017 (--)
----------
2018 (--)
----------
2019 (--)
----------
2020 (--)
----------
2021 (--)
----------
2022 (--)
----------
2023 (--)
----------
`;

exports[`<CalenderMonth /> tests dropdowns are updated when minDate, maxDate are given 3`] = `
----------
0 (-d)
----------
1 (-d)
----------
2 (--)
----------
3 (--)
----------
4 (--)
----------
5 (--)
----------
6 (--)
----------
7 (--)
----------
8 (--)
----------
9 (--)
----------
10 (--)
----------
11 (-d)
----------
`;

exports[`<CalenderMonth /> tests dropdowns are updated when minDate, maxDate are given 4`] = `
----------
2017 (--)
----------
`;

exports[`<CalenderMonth /> tests minDate and maxDate get priorty over startDate and endDate 1`] = `
------------------------------------------------------------------------------
|                                January 2017                                |
------------------------------------------------------------------------------
| 01 (--d) | 02 (--d) | 03 (--d) | 04 (--d) | 05 (--d) | 06 (--d) | 07 (--d) |
------------------------------------------------------------------------------
| 08 (--d) | 09 (--d) | 10 (--d) | 11 (---) | 12 (---) | 13 (---) | 14 (---) |
------------------------------------------------------------------------------
| 15 (---) | 16 (---) | 17 (---) | 18 (---) | 19 (---) | 20 (---) | 21 (---) |
------------------------------------------------------------------------------
| 22 (--d) | 23 (--d) | 24 (--d) | 25 (--d) | 26 (--d) | 27 (--d) | 28 (--d) |
------------------------------------------------------------------------------
| 29 (--d) | 30 (--d) | 31 (--d) | 01 (--d) | 02 (--d) | 03 (--d) | 04 (--d) |
------------------------------------------------------------------------------
| 05 (--d) | 06 (--d) | 07 (--d) | 08 (--d) | 09 (--d) | 10 (--d) | 11 (--d) |
------------------------------------------------------------------------------
`;

exports[`<CalenderMonth /> tests render prop has proper days selected when given 1`] = `
------------------------------------------------------------------------------
|                                January 2017                                |
------------------------------------------------------------------------------
| 01 (--d) | 02 (--d) | 03 (--d) | 04 (--d) | 05 (--d) | 06 (---) | 07 (---) |
------------------------------------------------------------------------------
| 08 (---) | 09 (---) | 10 (---) | 11 (sr-) | 12 (-r-) | 13 (-r-) | 14 (-r-) |
------------------------------------------------------------------------------
| 15 (-r-) | 16 (-r-) | 17 (-r-) | 18 (-r-) | 19 (-r-) | 20 (-r-) | 21 (sr-) |
------------------------------------------------------------------------------
| 22 (---) | 23 (---) | 24 (---) | 25 (---) | 26 (---) | 27 (--d) | 28 (--d) |
------------------------------------------------------------------------------
| 29 (--d) | 30 (--d) | 31 (--d) | 01 (--d) | 02 (--d) | 03 (--d) | 04 (--d) |
------------------------------------------------------------------------------
| 05 (--d) | 06 (--d) | 07 (--d) | 08 (--d) | 09 (--d) | 10 (--d) | 11 (--d) |
------------------------------------------------------------------------------
`;

exports[`<CalenderMonth /> tests showDropdowns gives apt props 1`] = `
----------
0 (--)
----------
1 (--)
----------
2 (--)
----------
3 (--)
----------
4 (--)
----------
5 (--)
----------
6 (--)
----------
7 (--)
----------
8 (--)
----------
9 (--)
----------
10 (--)
----------
11 (--)
----------
`;

exports[`<CalenderMonth /> tests showDropdowns gives apt props 2`] = `
----------
1969 (--)
----------
1970 (--)
----------
1971 (--)
----------
1972 (--)
----------
1973 (--)
----------
1974 (--)
----------
1975 (--)
----------
1976 (--)
----------
1977 (--)
----------
1978 (--)
----------
1979 (--)
----------
1980 (--)
----------
1981 (--)
----------
1982 (--)
----------
1983 (--)
----------
1984 (--)
----------
1985 (--)
----------
1986 (--)
----------
1987 (--)
----------
1988 (--)
----------
1989 (--)
----------
1990 (--)
----------
1991 (--)
----------
1992 (--)
----------
1993 (--)
----------
1994 (--)
----------
1995 (--)
----------
1996 (--)
----------
1997 (--)
----------
1998 (--)
----------
1999 (--)
----------
2000 (--)
----------
2001 (--)
----------
2002 (--)
----------
2003 (--)
----------
2004 (--)
----------
2005 (--)
----------
2006 (--)
----------
2007 (--)
----------
2008 (--)
----------
2009 (--)
----------
2010 (--)
----------
2011 (--)
----------
2012 (--)
----------
2013 (--)
----------
2014 (--)
----------
2015 (--)
----------
2016 (--)
----------
2017 (--)
----------
2018 (--)
----------
2019 (--)
----------
2020 (--)
----------
2021 (--)
----------
2022 (--)
----------
2023 (--)
----------
`;
